expr/functions/timeShift/function.go

```diff
@@ -89,31 +89,57 @@ func (f *timeShift) Do(ctx context.Context, e parser.Expr, from, until int64, va
 	if err != nil {
 		return nil, err
 	}
-	resetEndStr := strconv.FormatBool(resetEnd)
 
-	arg, err := helper.GetSeriesArg(ctx, e.Arg(0), from+int64(offs), until+int64(offs), values)
+	alignDST, err := e.GetBoolArgDefault(3, false)
+	if err != nil {
+		return nil, err
+	}
+
+	// Note: The fetch request is adjusted in expr.Metrics() to include both a request using the original
+	// start and stop time, and one that has the start and stop time adjusted based on the offset and alignDST, if relevant.
+	// This helps prevent having to re-fetch the data with the adjusted start and stop time from within this function.
+	arg, err := helper.GetSeriesArg(ctx, e.Arg(0), from, until, values)
 	if err != nil {
 		return nil, err
 	}
-	results := make([]*types.MetricData, len(arg))
 
-	for n, a := range arg {
+	newFrom := from + int64(offs)
+	newUntil := until + int64(offs)
+	if alignDST {
+		newFrom, newUntil, err = parser.AlignDST(from, until, offs, fconfig.Config.DefaultTimeZone)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	shiftedArgs, err := helper.GetSeriesArg(ctx, e.Arg(0), newFrom, newUntil, values)
+	if err != nil {
+		return nil, err
+	}
+
+	results := make([]*types.MetricData, len(shiftedArgs))
+
+	if len(arg) < 1 {
+		return []*types.MetricData{}, nil
+	}
+	series := arg[0]
+
+	for n, a := range shiftedArgs {
 		r := a.CopyLink()
-		r.Name = "timeShift(" + a.Name + ",'" + offsStr + "'," + resetEndStr + ")"
-		r.StartTime = a.StartTime - int64(offs)
-		r.StopTime = a.StopTime - int64(offs)
-		if resetEnd && r.StopTime > until {
-			r.StopTime = until
+		r.Name = "timeShift(" + a.Name + ",'" + offsStr + "'" + ")"
+		r.StartTime = series.StartTime
+		if resetEnd {
+			r.StopTime = series.StopTime
+		} else {
+			r.StopTime = a.StopTime - a.StartTime + series.StartTime
 		}
+		// Prevent using values that go past the stop time
 		length := int((r.StopTime - r.StartTime) / r.StepTime)
-		if length < 0 {
-			continue
+		if length >= 0 {
```

Comment:
This is what I meant [here ](https://github.com/grafana/carbonapi/pull/176#discussion_r1268644030)

```suggestion
		if length >= 0 && length < len(r.Values) {
```